portability-worker/src/main/java/org/dataportabilityproject/worker/WorkerMain.java

```diff
@@ -36,22 +40,43 @@ public static void main(String[] args) throws Exception {
 
         ExtensionContext extensionContext = getExtensionContext();
 
-        List<TransferExtension> extensions = new ArrayList<>();
-        ServiceLoader.load(TransferExtension.class).iterator().forEachRemaining(extensions::add);
+        Injector injector = Guice.createInjector(new WorkerModule());
+        WorkerImpl worker = injector.getInstance(WorkerImpl.class);
 
-        for (TransferExtension extension : extensions) {
-            extension.initialize(extensionContext);
+        // Poll job, and determine its data type, import service, and export service
+        worker.pollForJob();
+        WorkerJobMetadata metadata = worker.getWorkerJobMetadata();
+        String dataType = metadata.getDataType();
+        String exportService = metadata.getExportService();
+        String importService = metadata.getImportService();
+        Exporter<?, ?> exporter = null;
+        Importer<?, ?> importer = null;
+
+        // TODO: dynamically load the correct Importer & Exporter based on META-INF file name match.
+        for (TransferServiceProvider transferServiceProvider : listAllTransferServiceProviders()) {
+            if (transferServiceProvider.getServiceId().equals(exportService)) {
+                exporter = transferServiceProvider.getExporter(dataType);
+            } else if (transferServiceProvider.getServiceId().equals(importService)) {
+                importer = transferServiceProvider.getImporter(dataType);
+            }
         }
+        Preconditions.checkNotNull(exporter, String.format("No %s exporter found for %s",
+            dataType, exportService));
+        Preconditions.checkNotNull(importer, String.format("No %s importer found for %s",
+            dataType, importService));
+        // Create a TransferExtension with the correct importer and exporter, based on polled job
+        // TODO: load this as a single class rather than list, need to figure out the syntax
+        List<TransferExtension> extensions =
+            ImmutableList.of(createTransferExtension(exporter, importer));
```

Comment:
I'm a little confused why we're creating a new TransferExtension here. I thought the idea was that each service provider would create a TransferExtension which returns the importers and exporters that are valid for that service? Is that the wrong thought?